src/main/scala/chisel3/util/experimental/BoringUtils.scala

```diff
@@ -255,7 +256,7 @@ object BoringUtils {
             val bore = if (up) module.createSecretIO(purePortType) else module.createSecretIO(Flipped(purePortTypeBase))
             module.addSecretIO(bore)
             conLoc.asInstanceOf[RawModule].secretConnection(bore, rhs)
-            bore
+            bore.asInstanceOf[A]
```

Comment:
I'm not quite sure why I needed asInstanceOf here, and below. Without it, Scala gives me this error:

```
[error] /Users/mikeu/chisel/src/main/scala/chisel3/util/experimental/BoringUtils.scala:259:13: type mismatch;
[error]  found   : bore.type (with underlying type chisel3.BaseType)
[error]  required: A
[error]             bore
[error]             ^
```

I was under the impression that type `A` _is_ `BaseType`. Clearly I am out of my depth here... is this the right thing to do?

Reply:
what is the type of `bore`? It is the common ancestor of whatever `module.createSecretIO(purePortType)` and `module.createSecretIO(Flipped(purePortTypeBase)` is known to be... does createSecretIO properly return the types you want there?

Reply:
the compiler is complaining that while A is indeed a BaseType, it wants `bore` to be the specific subclass type `A`, not just a `BaseType`. something about that `if ... createSecretIO else ...` is making it lose the understanding that it is actually an `A`

Reply:
i think the problem is here, all it knows is that sometimes this is a Data and sometimes its an A so it concludes it can only be a BaseType:

Reply:
I think the problem is in this bit:
```
def purePortType = (createProbe, purePortTypeBase) match {
      case (Some(pi), d: Data) if pi.writable => RWProbe(d)
      case (Some(pi), d: Data) => Probe(d)
      case (_, _) => purePortTypeBase
    }
```

some of those cases are only Data, and some are your A type, the common ancestor of which is just BaseType. Can you make the A <: Data constraint be there instead of just Data?

Reply:
Thanks for the help! I will try it out.

Reply:
I tried poking around at this, and I'm still not sure the right way out of it. I'm not able to do `RWProbe(purePortTypeBase)` because you can only probe Data, which is why this change was introduced in the first place. I'm not sure how to express the constraint here.

Reply:
can you do:
```
      case (Some(pi), d: A <: Data) if pi.writable => RWProbe(d)
```

Reply:
Oh I see what you mean conceptually. I don't think this is legal scala syntax but i will see if i can express this.